src/phase0/state_transition/epoch_processing.rs

```diff
@@ -444,6 +509,350 @@ pub fn process_participation_record_updates<
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    context: &Context,
+) -> Vec<ValidatorIndex> {
+    let previous_epoch = get_previous_epoch(state, context);
+    state
+        .validators
+        .iter()
+        .enumerate()
+        .filter_map(|(i, v)| {
+            if is_active_validator(v, previous_epoch)
+                || (v.slashed && previous_epoch + 1 < v.withdrawable_epoch)
+            {
+                Some(i)
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<ValidatorIndex>>()
+}
+
+pub fn get_attestation_component_deltas<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+>(
+    state: &BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    attestations: &[PendingAttestation<MAX_VALIDATORS_PER_COMMITTEE>],
+    context: &Context,
+) -> Result<(Vec<Gwei>, Vec<Gwei>), Error> {
+    // Helper with shared logic for use by get source, target, and head deltas functions
+    let mut rewards = vec![0, state.validators.len() as Gwei];
+    let mut penalties = vec![0, state.validators.len() as Gwei];
+    let total_balance = get_total_active_balance(state, context)?;
+    let unslashed_attesting_indices =
+        get_unslashed_attesting_indices(state, attestations, context)?;
+    let attesting_balance = get_total_balance(state, &unslashed_attesting_indices, context)?;
+    for i in get_eligible_validator_indices(state, context) {
+        if unslashed_attesting_indices.contains(&i) {
+            let increment = context.effective_balance_increment; // Factored out from balance totals to avoid uint64 overflow
+            if is_in_inactivity_leak(state, context) {
+                // Since full base reward will be canceled out by inactivity penalty deltas,
+                // optimal participation receives full base reward compensation here.
+                rewards[i] += get_base_reward(state, i, context)?;
+            } else {
+                let reward_numerator =
+                    get_base_reward(state, i, context)? * (attesting_balance / increment);
+                rewards[i] += reward_numerator / (total_balance / increment);
+            }
+        } else {
+            penalties[i] += get_base_reward(state, i, context)?;
+        }
+    }
+    Ok((rewards, penalties))
```

Comment:
Also, random formatting preference -- return `Ok()`s with whitespace above it, or remove it (as it has been here)?

Reply:
likely context dependent :)

everyone will have their own preferences on formatting / layout -- i'd say just pick which one looks better to you.

my personal rule of thumb has to do w/ how related the return value is to the code before it -- if things are linked conceptually then i would group them closer together -- if the prior chunk of code is not for some reason then i'd put in some whitespace.... but not something to get super hung up on imo